pkg/virt-controller/watch/vm.go

```diff
@@ -448,21 +470,105 @@ func (c *VMController) startStop(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualM
 			return nil
 		}
 
+		log.Log.Object(vm).V(4).Info("Starting VMI")
 		err := c.startVMI(vm)
-		return err
-	}
+		if err != nil {
+			return err
+		}
+		return nil
+
+	case virtv1.RunStrategyRerunOnFailure:
+		// For this RunStrategy, a VMI should only be restarted if it failed.
+		// If a VMI enters the Succeeded phase, it should not be restarted.
+		if vmi != nil {
+			forceStop := false
+			// If there's a stop request that matches the existing VMI's UUID
+			if len(vm.Status.StateChangeRequests) != 0 {
+				stateChange := vm.Status.StateChangeRequests[0]
+				if stateChange.Action == virtv1.StopRequest &&
+					stateChange.UID != nil &&
+					*stateChange.UID == vmi.UID {
+					log.Log.Object(vm).V(4).Info("VMI should be stopped")
+					forceStop = true
+				}
+			}
+
+			if forceStop || vmi.Status.Phase == virtv1.Failed {
+				// For RerunOnFailure, this controller should only restart the VirtualMachineInstance
+				// if it failed.
+				log.Log.Object(vm).V(4).Info("Stopping VMI")
+				err := c.stopVMI(vm, vmi)
+				if err != nil {
+					log.Log.Object(vm).Errorf("Failure attempting to delete VMI: %v", err)
+					return err
+				}
+				// return to let the controller pick up the expected deletion
+			}
+			// VirtualMachineInstance is OK no need to do anything
+			return nil
+		}
+
+		log.Log.Object(vm).V(4).Info("Starting VMI")
+		err := c.startVMI(vm)
+		if err != nil {
+			return err
+		}
+		return nil
+
+	case virtv1.RunStrategyManual:
+		// For this RunStrategy, VMI's will be started/stopped/restarted using api endpoints only
+		if vmi != nil {
+			log.Log.Object(vm).V(4).Info("VMI exists")
+			forceStop := false
+			if len(vm.Status.StateChangeRequests) != 0 {
+				stateChange := vm.Status.StateChangeRequests[0]
+				if stateChange.Action == virtv1.StopRequest &&
+					stateChange.UID != nil &&
+					*stateChange.UID == vmi.UID {
+					log.Log.Object(vm).V(4).Info("VMI should be stopped")
+					forceStop = true
+				}
+			}
+			if forceStop || vmi.IsFinal() {
```

Comment:
Why do you try to delete a VMI which is in final state when no explicit stop was requested? Consider adding an issuer timestamp to the statechangerequest when doing starts and and compare it with the creation timestamp of the VMI. We can't just remove the VMI, which is the only reliable indicator right now on how the run of the VMI went.

Reply:
Ok fair point, but I don't understand what the timestamp would do. We could take the same approach for Manual that we do for RerunOnFailure--if a start request is initiated and a final-state VMI exists, treat the request as a restart. I'm concerned that adding a timestamp is going to trigger another moderate architecture change when it's not clear it's needed.

Reply:
Works for me.

Reply:
fixed